modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.catalog.serialization;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.apache.ignite.internal.util.ByteUtils;
+import org.apache.ignite.internal.util.io.IgniteDataInput;
+import org.apache.ignite.internal.util.io.IgniteDataOutput;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Utility methods used for serializing catalog objects.
+ */
+public class CatalogSerializationUtils {
+    /** Reads nullable string. */
+    public static @Nullable String readNullableString(DataInput in) throws IOException {
+        if (!in.readBoolean()) {
+            return null;
+        }
+
+        return in.readUTF();
+    }
+
+    /** Writes nullable string. */
+    public static void writeNullableString(@Nullable String str, DataOutput out) throws IOException {
+        out.writeBoolean(str != null);
+
+        if (str != null) {
+            out.writeUTF(str);
+        }
+    }
+
+    /** Writes collection containing strings. */
+    public static void writeStringCollection(Collection<String> list, DataOutput out) throws IOException {
+        out.writeInt(list.size());
+
+        for (String item : list) {
+            out.writeUTF(item);
+        }
+    }
+
+    /** Reads list of strings. */
+    public static List<String> readStringList(DataInput in) throws IOException {
```

Comment:
You can have a single method instead.
```suggestion
    public static <T extends Collection<String>> T readCollection(DataInput in, IntFunction<T> factory) throws IOException {
```

Reply:
Done, thanks